nixos/go-neb: init

mweinelt opened the pull request:
###### Motivation for this change

Go-neb is a matrix bot that was packaged in #87854. Now we deliver a simple module to easily set this up.



###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@tilpner @maralorn

mweinelt committed:
nixos/go-neb: init

aanderse commented on nixos/modules/services/networking/go-neb.nix:
```diff
@@ -0,0 +1,53 @@
+      wantedBy = [ "multi-user.target" ];
+
+      serviceConfig = {
+        Environment = [
```
Bonus points for using the [systemd.services.\<name\>.environment](https://nixos.org/nixos/options.html#systemd.+services.%3Cname%3E.environment) option.

mweinelt replied:
Sure, thanks for the pointer.

aanderse commented on nixos/modules/services/networking/go-neb.nix:
```diff
@@ -0,0 +1,53 @@
+      description = "Public-facing endpoint that can receive webhooks.";
+    };
+
+    config = mkOption {
```
Any interest in providing a mechanism for handling secrets in here? From the looks of the format likely not trivial to add great support...

mweinelt replied:
The only secret worth protecting would be `clients.*.AccessToken`. I'm not sure how I would keep that out of the nix store when the program does not support reading secrets from dedicated files.

aanderse replied:
Yeah the solution wouldn't be pretty. Some modules will read the password on startup and then copy+paste+modify the generated configuration file replacing any secrets in a protected directory. Using `DynamicUser` adds additional complications. Secrets management in NixOS is not an easy thing, unfortunately.

aanderse reviewed (commented):
Nice and simple module :+1:

Even more bonus points if it comes with a NixOS test :sparkles:

maralorn commented:
I have copied this module to my server and have alertmanager warnings (and a wikipedia bot) up and running with this.
```nix
{
  services.go-neb = {
    enable = true;
    baseUrl = "http://localhost";
    config = {
      clients = [{
        UserId = "@marabot:maralorn.de";
        AccessToken = config.m-0.private.matrix_marabot_token;
        HomeServerUrl = "https://matrix.maralorn.de";
        Sync = true;
        AutoJoinRooms = true;
        DisplayName = "marabot";
      }];
      realms = [ ];
      sessions = [ ];
      services = [
        {
          ID = "alertmanager_service";
          Type = "alertmanager";
          UserId = "@marabot:maralorn.de";
          Config = {
            webhook_url =
              "http://localhost:4050/services/hooks/YWxlcnRtYW5hZ2VyX3NlcnZpY2UK";
            rooms = {
              "!negVsngnYOmXYCLKiO:maralorn.de" = {
                text_template = ''
                  {{range .Alerts -}} [{{ .Status }}] {{index .Labels "alertname" }}: {{index .Annotations "description"}} {{ end -}}'';
                  html_template = ''
                    {{range .Alerts -}}
                      {{ $severity := index .Labels "severity" }}
                      {{ if eq .Status "firing" }}
                        {{ if eq $severity "critical"}}
                          <font color='red'><b>[FIRING - CRITICAL]</b></font>
                        {{ else if eq $severity "warning"}}
                          <font color='orange'><b>[FIRING - WARNING]</b></font>
                        {{ else }}
                          <b>[FIRING - {{ $severity }}]</b>
                        {{ end }}
                      {{ else }}
                        <font color='green'><b>[RESOLVED]</b></font>
                      {{ end }}
                      {{ index .Labels "alertname"}}:{{ index .Annotations "description"}}
                      <a href="{{ .GeneratorURL }}">source</a><br/>
                    {{end -}}
                    '';
                msg_type = "m.text"; # Must be either `m.text` or `m.notice`
              };
            };
          };
        }
        {
          ID = "wikipedia_service";
          Type = "wikipedia";
          UserID = "@marabot:maralorn.de"; # requires a Syncing client
          Config = { };
        }
      ];
    };
  };
}
```

mweinelt committed:
nixos/tests/go-neb: init

Simple test to start the unit with a dummy client and a dummy service
and check if it can receive webhooks.

mweinelt commented:
Simple test added.

aanderse commented on nixos/modules/services/networking/go-neb.nix:
```diff
@@ -0,0 +1,53 @@
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+      environment = {
+        BASE_URL = "${cfg.baseUrl}";
```
Quotes no longer required here.

mweinelt replied:
Indeed!

tilpner commented:
In https://github.com/matrix-org/go-neb#quick-start, DATABASE_{TYPE,URL} are set, which is not replicated here. Where does this module store its state?

tilpner commented on nixos/modules/services/networking/go-neb.nix:
```diff
@@ -0,0 +1,53 @@
+    };
+
+    config = mkOption {
+      type = types.attrs;
```
Multiple definitions of attrs are apparently not merged properly. [According to @Infinisil, `types.uniq types.attr` might be appropriate](https://logs.nix.samueldr.com/nixos-dev/2020-06-02#3549491;) to prevent that merging.

maralorn replied:
That mistake is probably all over nixpkgs right? But seems easy enough to fix here.

mweinelt replied:
In fact i don't see that used anywhere. Wild!

maralorn commented:
@tilpner the bot has two modes. Statefully saving the config in an sqlite DB or reading the config from a config file. So the DB you are mentioning is not data state but only for config state. We opted to support only the second option, because it supports the same features and is much more nix like.

mweinelt commented:
> @tilpner the bot has two modes. Statefully saving the config in an sqlite DB or reading the config from a config file. So the DB you are mentioning is not data state but only for config state. We opted to support only the second option, because it supports the same features and is much more nix like.

The partial source for this statement is here: https://github.com/matrix-org/go-neb/blob/81842c69f831b7dccaa685ccacbbe218d059d76c/config.sample.yaml#L5-L9

tilpner commented:
I see. It surprised me that they were exclusive modes, I was hoping it would operate in a hybrid mode, as e.g. [reliable RSS feed consumption requires state](https://github.com/matrix-org/go-neb/blob/81842c69f831b7dccaa685ccacbbe218d059d76c/services/rssbot/rssbot.go#L97-L98).

Supporting only the declarative mode in this module is reasonable though. Thanks for writing it! :)

maralorn commented:
@aanderse Are you also satisfied with the state of this?

aanderse reviewed (approved):
Yeah this is a winner in my books. Thanks for the extra work on this one. Bonus points for all! :sparkles:

I'm not a subject matter expert so I wouldn't think to merge this... but if you're just waiting for a merge and everyone agrees I'm certainly happy to do so based on that :+1:

aanderse commented:
... and just as I say that I notice the test isn't added to the list of tests. Please add and we're good.

mweinelt commented:
Good catch. Added here: https://github.com/NixOS/nixpkgs/pull/89327/files#diff-8e2944fa8b75f34d13428287028e1323R122

Thanks!

maralorn commented:
@aanderse Can you maybe point us to a subject matter expert? Otherwise I'd opt for the merge.

aanderse commented:
@maralorn maybe @Ma27 knows a thing or two... but this looks really good, so I'll merge by end of day if no one has any objections.

Thanks! :tada:

aanderse commented:
@GrahamcOfBorg test go-neb

aanderse commented:
Oops! Thanks team :heart:
